helix-core/src/main/java/org/apache/helix/util/MessageUtil.java

```diff
@@ -0,0 +1,134 @@
+package org.apache.helix.util;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.UUID;
+
+import org.apache.helix.HelixManager;
+import org.apache.helix.model.LiveInstance;
+import org.apache.helix.model.Message;
+import org.apache.helix.model.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Message utils to operate on message such creating messages.
+ */
+public class MessageUtil {
```

Comment:
I think it would be a more elegant design if the logic in this util is done as MessageBuilder. Mainly to avoid long parameter lists and easier for future extension.

Reply:
I thought about it. The advantage of using builder pattern is it supports different combinations of the arguments.
However, for the message creation methods, a fixed number of arguments/fields are required, eg:
```
Resource resource, String partitionName, String instanceName, String sessionId,
      String stateModelDefName, String fromState, String toState, String nextState
```
Using the method arguments gives a reminder that we need to pass in the arguments, which builder pattern does not. I can also imagine that it'll need more code to validate the arguments in the builder. Using builder pattern does not give better controller of construction for this case here. And it would also have a long list of setMethod():
```
newBuilder()
.setA()
.setB()
.setC()
...
.setX()
```
So I chose to keep the original long list of arguments.

Reply:
The builder also helps to avoid too many constructors if we keep adding parameters and there are different combinations. But I don't think it is a major concern here. Let's keep the current implementation.